lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -69,54 +128,18 @@ class AllocaPattern : public OpRewritePattern<cudaq::cc::AllocaOp> {
       auto ptrTy = cudaq::cc::PointerType::get(arrTy);
       // Build a new name based on the kernel name.
       std::string name = funcName + ".rodata_" + std::to_string(counter++);
-      {
-        OpBuilder::InsertionGuard guard(rewriter);
-        if (auto complexTy = dyn_cast<ComplexType>(eleTy)) {
-          // Transforming complex vectors is a bit more labor intensive. Use the
-          // IRBuilder to create the object since we have to thread the needle
-          // for the LLVM-IR to be lowered to LLVM correctly.
-          auto transform = [&]<typename A>(SmallVectorImpl<Attribute> &vec)
-              -> std::vector<std::complex<A>> {
-            std::vector<std::complex<A>> result;
-            for (auto a : vec) {
-              auto v = cast<ArrayAttr>(a);
-              if constexpr (std::is_same_v<A, double>) {
-                result.emplace_back(
-                    cast<FloatAttr>(v[0]).getValue().convertToDouble(),
-                    cast<FloatAttr>(v[1]).getValue().convertToDouble());
-              } else {
-                result.emplace_back(
-                    cast<FloatAttr>(v[0]).getValue().convertToFloat(),
-                    cast<FloatAttr>(v[1]).getValue().convertToFloat());
-              }
-            }
-            return result;
-          };
-          cudaq::IRBuilder irBuilder(rewriter.getContext());
-          if (complexTy.getElementType() == rewriter.getF64Type()) {
-            std::vector<std::complex<double>> vals =
-                transform.template operator()<double>(values);
-            irBuilder.genVectorOfComplexConstant(loc, module, name, vals);
-          } else {
-            std::vector<std::complex<float>> vals =
-                transform.template operator()<float>(values);
-            irBuilder.genVectorOfComplexConstant(loc, module, name, vals);
-          }
-        } else {
-          OpBuilder::InsertionGuard guard(rewriter);
-          rewriter.setInsertionPointToEnd(module.getBody());
-          rewriter.create<cudaq::cc::GlobalOp>(loc, arrTy, name, valuesAttr,
-                                               /*isConstant=*/true,
-                                               /*isExternal=*/false);
-        }
-      }
+      cudaq::IRBuilder irBuilder(rewriter.getContext());
+      genVectorOfConstantsFromAttributes(irBuilder, loc, module, name, values,
```

Comment:
I'm not sure I like this change or not.

This code was converting all the vectors, regardless of type to globals of vector of complex of double here at compile time. That costs compilation time, some memory, and some disk/file space for the executable, etc.

It looks like these changes do not do that conversion, but kick the can down the road. So we can end up with a family of vectors of all sorts of floating point types. That may save a some disk space, but every time the executable runs, those arrays will need to be converted to the API of the runtime, which I believe is vector of complex of double.

Reply:
Not sure I understand - I just moved the code into helpers and used a common API (plus made vectors of float treated the same way as vectors of complex). I don't think any ordering should have changed, please correct me if I am wrong.

Reply:
Don't the helpers create one of `vector<float>`, `vector<double>`, `vector<complex<float>>`, etc.?

Previously this created only `vector<complex<double>>` and the other possibilities did not exist intentionally.

Reply:
I updated to only create globals if we see an argument flowing into a` quake.state_init`, so nothing should change for previously handled code